app/code/core/Mage/Catalog/Model/Product.php

```diff
@@ -581,23 +581,21 @@ public function getCategory()
     /**
      * Set assigned category IDs array to product
      *
-     * @param array|string $ids
+     * @param array|int|string $ids
      * @return $this
      */
     public function setCategoryIds($ids)
     {
-        if (is_string($ids)) {
+        if (is_string($ids) === true) {
             $ids = explode(',', $ids);
-        } elseif (!is_array($ids)) {
+            $ids = array_map('trim', $ids);
+        } elseif (is_int($ids) === true) {
+            $ids = (array) $ids;
+        } elseif (is_array($ids) === false) {
             Mage::throwException(Mage::helper('catalog')->__('Invalid category IDs.'));
         }
-        foreach ($ids as $i => $v) {
-            if (empty($v)) {
-                unset($ids[$i]);
-            }
-        }
 
-        $this->setData('category_ids', $ids);
+        $this->setData('category_ids', array_filter($ids));
```

Comment:
`array_map('intval', $ids)` here and then filter.

Reply:
Good for the string option,  yes